File: src/noise.ts
export interface FbmOptions {
  octaves?: number;
  persistence?: number;
  lacunarity?: number;
  frequency?: number;
  amplitude?: number;
}

// Quintic fade: 6t^5 - 15t^4 + 10t^3
function fade(t: number): number {
  return t * t * t * (t * (t * 6 - 15) + 10);
}

function lerp(a: number, b: number, t: number): number {
  return a + t * (b - a);
}

// 2D gradient vectors (8 directions)
const GRAD2_X = [1, -1, 1, -1, 1, -1, 0, 0];
const GRAD2_Z = [0, 0, 1, 1, -1, -1, 1, -1];

function grad2D(hash: number, x: number, z: number): number {
  const h = hash & 7;
  return GRAD2_X[h] * x + GRAD2_Z[h] * z;
}

export class PerlinNoise {
  private perm: Uint8Array;

  constructor(seed: number = 42) {
    // Build seeded permutation table
    const p = new Uint8Array(256);
    for (let i = 0; i < 256; i++) p[i] = i;

    // Fisher-Yates shuffle with simple seedable PRNG (mulberry32)
    let s = seed | 0;
    for (let i = 255; i > 0; i--) {
      // mulberry32 step
      s = (s + 0x6d2b79f5) | 0;
      let t = Math.imul(s ^ (s >>> 15), 1 | s);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      const r = ((t ^ (t >>> 14)) >>> 0) % (i + 1);
      const tmp = p[i];
      p[i] = p[r];
      p[r] = tmp;
    }

    // Double the table to avoid index wrapping
    this.perm = new Uint8Array(512);
    for (let i = 0; i < 512; i++) this.perm[i] = p[i & 255];
  }

  noise2D(x: number, z: number): number {
    const perm = this.perm;

    // Grid cell coordinates
    const xi = Math.floor(x) & 255;
    const zi = Math.floor(z) & 255;

    // Fractional position within cell
    const xf = x - Math.floor(x);
    const zf = z - Math.floor(z);

    // Fade curves
    const u = fade(xf);
    const v = fade(zf);

    // Hash corners
    const aa = perm[perm[xi] + zi];
    const ab = perm[perm[xi] + zi + 1];
    const ba = perm[perm[xi + 1] + zi];
    const bb = perm[perm[xi + 1] + zi + 1];

    // Gradient dot products at corners, bilinear interpolation
    return lerp(
      lerp(grad2D(aa, xf, zf), grad2D(ba, xf - 1, zf), u),
      lerp(grad2D(ab, xf, zf - 1), grad2D(bb, xf - 1, zf - 1), u),
      v,
    );
  }

  fbm2D(x: number, z: number, opts?: FbmOptions): number {
    const octaves = opts?.octaves ?? 6;
    const persistence = opts?.persistence ?? 0.5;
    const lacunarity = opts?.lacunarity ?? 2.0;
    let freq = opts?.frequency ?? 0.01;
    let amp = opts?.amplitude ?? 1.0;

    let sum = 0;
    let maxAmp = 0;

    for (let i = 0; i < octaves; i++) {
      sum += this.noise2D(x * freq, z * freq) * amp;
      maxAmp += amp;
      amp *= persistence;
      freq *= lacunarity;
    }

    return sum / maxAmp; // Normalize to [-1, 1]
  }
}

File: server/voxel_server/noise.py
"""Vectorized 2D Perlin noise with FBM, matching the frontend PerlinNoise class."""

from __future__ import annotations

import numpy as np


# 8 gradient directions (matching frontend GRAD2_X/GRAD2_Z)
_GRAD2_X = np.array([1, -1, 1, -1, 1, -1, 0, 0], dtype=np.float64)
_GRAD2_Z = np.array([0, 0, 1, 1, -1, -1, 1, -1], dtype=np.float64)


def _build_perm(seed: int) -> np.ndarray:
    """Build a 512-entry permutation table using mulberry32 PRNG (matches frontend)."""
    p = np.arange(256, dtype=np.int32)

    # Fisher-Yates shuffle with mulberry32 (must match JS exactly)
    s = np.int32(seed)
    for i in range(255, 0, -1):
        # mulberry32 step (matching JS Math.imul and unsigned shifts)
        s = np.int32(np.int64(s) + np.int64(0x6D2B79F5))
        t = np.int64(s ^ (np.uint32(s) >> np.uint32(15))) * np.int64(1 | s)
        t = np.int64(np.int32(t))
        t2 = np.int64(np.int32(t ^ (np.uint32(np.int32(t)) >> np.uint32(7))))
        t2 = np.int64(np.int32(t2 * np.int64(61 | np.int32(t2)))) ^ t2
        r_val = int(np.uint32(np.int32(t2) ^ (np.uint32(np.int32(t2)) >> np.uint32(14))))
        r = r_val % (i + 1)
        p[i], p[r] = p[r], p[i]

    perm = np.empty(512, dtype=np.int32)
    perm[:256] = p
    perm[256:] = p
    return perm


def _fade(t: np.ndarray) -> np.ndarray:
    """Quintic fade: 6t^5 - 15t^4 + 10t^3"""
    return t * t * t * (t * (t * 6 - 15) + 10)


def _grad2d(hashes: np.ndarray, dx: np.ndarray, dz: np.ndarray) -> np.ndarray:
    """Vectorized gradient dot product."""
    h = hashes & 7
    return _GRAD2_X[h] * dx + _GRAD2_Z[h] * dz


class PerlinNoise:
    """2D Perlin noise generator with seeded permutation table."""

    def __init__(self, seed: int = 42) -> None:
        self.perm = _build_perm(seed)

    def noise2d(self, x: np.ndarray, z: np.ndarray) -> np.ndarray:
        """Evaluate 2D Perlin noise at arrays of (x, z) coordinates.

        x, z should be broadcastable numpy arrays. Returns array of noise values in ~[-1, 1].
        """
        perm = self.perm

        xi = np.floor(x).astype(np.int32) & 255
        zi = np.floor(z).astype(np.int32) & 255

        xf = x - np.floor(x)
        zf = z - np.floor(z)

        u = _fade(xf)
        v = _fade(zf)

        # Hash corners
        aa = perm[perm[xi] + zi]
        ab = perm[perm[xi] + zi + 1]
        ba = perm[perm[xi + 1] + zi]
        bb = perm[perm[xi + 1] + zi + 1]

        # Gradient dot products + bilinear interpolation
        x1 = _grad2d(aa, xf, zf) * (1 - u) + _grad2d(ba, xf - 1, zf) * u
        x2 = _grad2d(ab, xf, zf - 1) * (1 - u) + _grad2d(bb, xf - 1, zf - 1) * u

        return x1 * (1 - v) + x2 * v

    def fbm2d(
        self,
        x: np.ndarray,
        z: np.ndarray,
        *,
        frequency: float = 0.006,
        octaves: int = 6,
        persistence: float = 0.4,
        lacunarity: float = 2.0,
        amplitude: float = 1.0,
    ) -> np.ndarray:
        """Fractal Brownian Motion using Perlin noise."""
        result = np.zeros_like(x, dtype=np.float64)
        max_amp = 0.0
        freq = frequency
        amp = amplitude

        for _ in range(octaves):
            result += self.noise2d(x * freq, z * freq) * amp
            max_amp += amp
            amp *= persistence
            freq *= lacunarity

        return result / max_amp
